zkvm/src/types.rs

```diff
@@ -281,15 +317,25 @@ impl Data {
     pub fn len(&self) -> usize {
         match self {
             Data::Opaque(data) => data.len(),
+            // TBD: calculate length of witness data
+            // Is there a way to do this without just serializing the
+            // data and taking its length?
```

Comment:
We can have `serialized_len()` method on all types that can be captured by `Data` and call them recursively to add them up. Not sure how tedious it would be.

Alternative is to rename this method to `minimum_serialized_length()` which guarantees to give a lower bound, which is what we need for the Vec allocation. And then return 0 for Data::Witness, which means the prover would have to re-allocate the Vec as it builds it, but it's good for now and probably will never matter.

Reply:
Although we need to check whether the .len() is used where exact length is expected. Then we have to scratch our heads more.